gateway-service-metadata/src/main/java/org/apache/knox/gateway/service/metadata/ServiceModel.java

```diff
@@ -0,0 +1,121 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.knox.gateway.service.metadata;
+
+import java.util.Locale;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+
+import org.apache.knox.gateway.service.definition.Metadata;
+import org.apache.knox.gateway.topology.Service;
+
+@XmlRootElement(name = "service")
+@XmlAccessorType(XmlAccessType.NONE)
+public class ServiceModel {
+
+  public enum Type {
+    API, UI, UNKNOWN
+  };
+
+  protected HttpServletRequest request;
+  protected String topologyName;
+  protected String gatewayPath;
+  protected Service service;
+  protected Metadata serviceMetadata;
+
+  public void setRequest(HttpServletRequest request) {
+    this.request = request;
+  }
+
+  public void setTopologyName(String topologyName) {
+    this.topologyName = topologyName;
+  }
+
+  public void setGatewayPath(String gatewayPath) {
+    this.gatewayPath = gatewayPath;
+  }
+
+  public void setService(Service service) {
+    this.service = service;
+  }
+
+  public void setServiceMetadata(Metadata serviceMetadata) {
+    this.serviceMetadata = serviceMetadata;
+  }
+
+  @XmlElement
+  public String getServiceName() {
+    return this.service == null ? "" : service.getRole();
+  }
+
+  @XmlElement
+  public String getVersion() {
+    return this.service == null ? "" : this.service.getVersion() == null ? "" : this.service.getVersion().toString();
+  }
+
+  @XmlElement(name = "shortDesc")
+  public String getShortDescription() {
+    if (serviceMetadata == null) {
+      return getServiceName().substring(0, 1).toUpperCase(Locale.ROOT) + getServiceName().substring(1).toLowerCase(Locale.ROOT);
+    } else {
+      return serviceMetadata.getShortDesc();
+    }
+  }
+
+  @XmlElement
+  public String getDescription() {
+    if (serviceMetadata == null) {
+      return getShortDescription() + (Type.API == getType() ? " REST API" : Type.UI == getType() ? " Web User Interface" : "");
+    } else {
+      return serviceMetadata.getDescription();
+    }
+  }
+
+  @XmlElement
+  public Type getType() {
+    return serviceMetadata == null ? Type.UNKNOWN : Type.valueOf(serviceMetadata.getType());
+  }
+
+  @XmlElement
+  public String getContext() {
+    return (serviceMetadata == null ? "/" + getServiceName().toLowerCase(Locale.ROOT) : serviceMetadata.getContext()) + "/";
+  }
+
+  @XmlElement
+  public String getServiceUrl() {
+    String context = getContext();
+    if ("HIVE".equals(getServiceName())) {
```

Comment:
Long term I hope we don't have to do this in the ServiceModel? It would be better to have this captured in the service metadata.